gateway/src/main.rs

```diff
@@ -0,0 +1,314 @@
+// TODO: ~~API crate should expose the active deployed port for a service~~
+// TODO: ~~gateway crate should poll active deployment port for proxy~~
+// TODO: ~~gateway crate should rewrite the projects -> services route~~
+// TODO: client should create project then push new deployment (refactor endpoint)
+// TODO: ~~rename API crate~~
+// TODO: ~~move common things to the common crate~~
+// TODO: ~~AccountName and ProjectName validation logic?~~
+// TODO: Add some tests (ideas?)
+// TODO: Implement the delete project endpoint to make sure users can
+//       self-serve out of issues
+// TODO: ~~Do a `docker pull` of the target runtime image to use when
+//       starting up~~
+
+#![allow(warnings)]
+
+#[macro_use]
+extern crate async_trait;
+
+#[macro_use]
+extern crate log;
+extern crate core;
+
+use std::error::Error as StdError;
+use std::fmt::Formatter;
+use std::io;
+use std::net::IpAddr;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::convert::Infallible;
+
+use axum::Json;
+use axum::http::StatusCode;
+use axum::response::{
+    IntoResponse,
+    Response
+};
+use bollard::Docker;
+use convert_case::{Casing, Case};
+use serde::{
+    Deserialize,
+    Deserializer,
+    Serialize
+};
+use serde_json::json;
+use clap::Parser;
+
+use crate::api::make_api;
+use crate::proxy::make_proxy;
+use crate::service::GatewayService;
+use crate::args::Args;
+
+pub mod api;
+pub mod project;
+pub mod proxy;
+pub mod service;
+pub mod auth;
+pub mod args;
+
+#[derive(Debug)]
+pub enum ErrorKind {
+    KeyMissing,
+    BadHost,
+    KeyMalformed,
+    Unauthorized,
+    UserNotFound,
+    ProjectNotFound,
+    InvalidProjectName,
+    ProjectAlreadyExists,
+    ProjectNotReady,
+    ProjectUnavailable,
+    InvalidOperation,
+    Internal
+}
+
+impl std::fmt::Display for ErrorKind {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let formatted = format!("{:?}", self).to_case(Case::Snake);
+        write!(f, "{}", formatted)
+    }
+}
+
+/// Server-side errors that do not have to do with the user runtime
+/// should be [`Error`]s.
+///
+/// All [`Error`] have an [`ErrorKind`] and an (optional) source.
+
+/// [`Error] is safe to be used as error variants to axum endpoints
+/// return types as their [`IntoResponse`] implementation does not
+/// leak any sensitive information.
+#[derive(Debug)]
+pub struct Error {
+    kind: ErrorKind,
+    source: Option<Box<dyn StdError + Sync + Send + 'static>>
+}
+
+impl Error {
+    pub fn source<E: StdError + Sync + Send + 'static>(kind: ErrorKind, err: E) -> Self {
+        Self {
+            kind,
+            source: Some(Box::new(err))
+        }
+    }
+
+    pub fn custom<S: AsRef<str>>(kind: ErrorKind, message: S) -> Self {
+        Self {
+            kind,
+            source: Some(Box::new(io::Error::new(io::ErrorKind::Other, message.as_ref().to_string())))
+        }
+    }
+
+    pub fn kind(kind: ErrorKind) -> Self {
```

Comment:
I've noticed inconsistent use  of `kind` and `From<ErrorKind::from` throughout the codebase - might be worth removing `kind` and just using `from` for consistency and out-of-the-box compatibility.

Reply:
I'll refactor the `From` to dispatch through `Error::kind`.